pkg/kexec/memory_linux.go

```diff
@@ -0,0 +1,461 @@
+
+		const (
+			// file names
+			start = "start"
+			end   = "end"
+			typ   = "type"
+		)
+
+		base := path.Base(name)
+		if base != start && base != end && base != typ {
+			return fmt.Errorf("unexpected file %q", name)
+		}
+		dir := path.Dir(name)
+
+		b, err := ioutil.ReadFile(name)
+		if err != nil {
+			return fmt.Errorf("error reading file %q: %v", name, err)
+		}
+
+		data := strings.TrimSpace(string(b))
+		r := ranges[dir]
+		if base == typ {
+			r.typ = RangeType(data)
+			ranges[dir] = r
+			return nil
+		}
+
+		v, err := strconv.ParseUint(data, 0, 64)
+		if err != nil {
+			return err
+		}
+		switch base {
+		case start:
+			r.start = uintptr(v)
+		case end:
+			r.end = uintptr(v)
+		}
+		ranges[dir] = r
+		return nil
+	}
+
+	if err := filepath.Walk(memoryMapRoot, walker); err != nil {
+		return err
+	}
+
+	for _, r := range ranges {
+		m.Phys = append(m.Phys, TypedAddressRange{
+			Range: Range{
+				Start: r.start,
+				Size:  uint(r.end - r.start),
+			},
+			Type: r.typ,
+		})
+	}
+	sort.Slice(m.Phys, func(i, j int) bool {
+		return m.Phys[i].Start < m.Phys[j].Start
+	})
+	return nil
+}
+
+// RangeType defines type of a TypedAddressRange based on the Linux
+// kernel string provided by firmware memory map.
+type RangeType string
+
+const (
+	RangeRAM     RangeType = "System RAM"
+	RangeDefault           = "Default"
+	RangeNVACPI            = "ACPI Non-volatile Storage"
+	RangeACPI              = "ACPI Tables"
+	RangeNVS               = "Reserved"
+)
+
+// Memory provides routines to work with physical memory ranges.
+type Memory struct {
+	Phys []TypedAddressRange
+
+	Segments []Segment
+}
+
+// PhyscalMemory represents range of physical memory.
+type TypedAddressRange struct {
+	Range
+	Type RangeType
+}
+
+var ErrNotEnoughSpace = errors.New("not enough space")
+
+// FindSpace returns pointer to the physical memory,
+// where array of size sz can be stored during next
+// AddKexecSegment call.
+func (m Memory) FindSpace(sz uint) (start uintptr, err error) {
+	pageSize := uint(os.Getpagesize())
+	ranges := m.sub()
+	sz = (sz + pageSize - 1) &^ (pageSize - 1)
+	for _, r := range ranges {
+		if r.Size >= sz {
+			return r.Start, nil
+		}
+	}
+	return 0, ErrNotEnoughSpace
+}
+
+func (m *Memory) addKexecSegment(addr uintptr, d []byte) {
+	s := NewSegment(d, Range{
+		Start: addr,
+		Size:  uint(len(d)),
+	})
+	s = AlignPhys(s)
+	m.Segments = append(m.Segments, s)
+	sort.Slice(m.Segments, func(i, j int) bool {
+		return m.Segments[i].Phys.Start < m.Segments[j].Phys.Start
+	})
+}
+
+// AddKexecSegment adds d to a new kexec segment
+func (m *Memory) AddKexecSegment(d []byte) (addr uintptr, err error) {
+	size := uint(len(d))
+	start, err := m.FindSpace(size)
+	if err != nil {
+		return 0, err
+	}
+	m.addKexecSegment(start, d)
+	return start, nil
+}
+
+// sub subtracts physycal ranges of kexec segments from
+// RAM segments of TypedAddressRange.
+//
+// E.g if kexec segments are
+//            [{start:0 size:100} {start:200 size:100}]
```

Comment:
This example doesn't make much sense, since kexec segments should be a subset of ram segments, no?

Reply:
Fixed the comment. If they are not the subset of RAM segments then it won't work. In this case in theory we can do some relocations, but I don't want to think about it now.